### PR TITLE
KLIP-19: Introduce Materialized Views

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

Initial drop of KLIP-19, which introduces Materialized Views to KSQL.

Note: there is some overlap between this KLIP and [KLIP-11](https://github.com/confluentinc/ksql/pull/3799). However, I think there is benefit in pulling out and expanding on the introduction of `MATERIALIZED VIEW` into its own KLIP,  where it can be discussed and decided upon independently. If/when this KLIP is approved, KLIP-11 can be updated, making it more targeted.

### CODE REVIEW

File: design-proposals/klip-19-materialize-views.md [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+At the moment, users can create a new table in KSQL using a `CREATE TABLE` statement with `PARTITIONS` and, optionally,
+`REPLICAS`. This will create a new topic. As no other system writes to this topic, KSQL can support `INSERT`s into it.
+Likewise, a `CREATE TABLE` statement can be used to import an existing topic into KSQL. As long as no other system is 
+writing to that topic, (and that includes another KSQL topology), KSQL can offer correct `INSERT` semantics.
+
+The defining factor here is that it is only possible to have correct `INSERT` semantics _if no other system updates the table_.
+
+**Side note**: It would be easy to support correct `DELETE` semantics too. However, correct `UPDATE` row semantics will
+require more work. Likely, it will involve proxying calls to the KSQL node that owns the key, ensuring a single writer
+pattern.
```
Comment (mjsax) [inline design-proposals/klip-19-materialize-views.md]: Why is that required?

File: design-proposals/klip-19-materialize-views.md [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+To ensure correct table semantics KSQL needs to differentiate between sources where it is the sole writer and those updated
+by other systems. Tables where KSQL is the only writer are _source datasets_. Topic containing data updated by another system, 
+including another KSQL query, should be treated as _read-only_ as KSQL can not offer correct table semantics on `INSERT`s.
+
+SQL already has concepts to differentiate: 
+ * a `TABLE` is a source dataset. The data in the table can be mutated via `INSERT`/`UPDATE`/`DELETE`s.
+ * a `VIEW` is a derived and read-only dataset. It's state is _derived_ from one or more sources. 
+  `INSERT`/`UPDATE`/`DELETE`s are not supported. 
+ * A `MATERIALIZED VIEW` is a `VIEW` where the cost of calculating the derived dataset is done on the write path, rather 
+   than the read.
```
Comment (mjsax) [inline design-proposals/klip-19-materialize-views.md]: I think we need to clarify what `MATERIALIZED` implies, because in ksqlDB (in contrast to a RDBMS) it could mean to write into a topic (output or changelog?) and/or state store.

Given the remainder of the KLIP, I assume you propose that `MATERIALIZED` only implies that there will be an output topic, and it does _not_ imply anything with regard to state stores? I am not 100% sure atm, especially given "pull queries", if those semantics will be confusion to users (ie, "it's materialized, why can't I query it?" or later "it's materialized, why are my queries so slow"). To be fair, I don't have a better proposal atm, but want to raise this point as it seems important.

File: design-proposals/klip-19-materialize-views.md [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+* Add `CREATE MATERIALIZED STREAM VIEW` to import an existing topic containing a stream.
+    * `INSERT`s will *not* be allowed into the view, as views are read-only.
+    * `PARTITIONS` and `REPLICAS` *not* supported in `WITH` clause.
+* Re-purpose `CREATE STREAM` to create a mutable stream backed by a new or existing topic.
+    * `INSERT`s into the stream will be supported.
+    * `PARTITIONS` and `REPLICAS` supported in `WITH` clause to control customisation of any new topic. 
+* Replace `CREATE STREAM AS SELECT` with `CREATE MATERIALIZED STREAM VIEW AS SELECT`.
+    * `INSERT`s will *not* be allowed into the view, as views are read-only.
+    * `PARTITIONS` and `REPLICAS` supported in `WITH` clause to control customisation of the sink topic.   
+* Either `CREATE STREAM AS SELECT` will be removed, or maintained as creating a mutable derived stream.
```
Comment (mjsax) [inline design-proposals/klip-19-materialize-views.md]: IMHO, keeping `CSAS` is useful.

File: design-proposals/klip-19-materialize-views.md [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+    * `INSERT`s into the stream will be supported.
+    * `PARTITIONS` and `REPLICAS` supported in `WITH` clause to control customisation of any new topic. 
+* Replace `CREATE STREAM AS SELECT` with `CREATE MATERIALIZED STREAM VIEW AS SELECT`.
+    * `INSERT`s will *not* be allowed into the view, as views are read-only.
+    * `PARTITIONS` and `REPLICAS` supported in `WITH` clause to control customisation of the sink topic.   
+* Either `CREATE STREAM AS SELECT` will be removed, or maintained as creating a mutable derived stream. 
+    * `INSERT`s into the stream will be supported.
+    * `PARTITIONS` and `REPLICAS` supported in `WITH` clause to control customisation of the sink topic.
+    
+For those worried about the excessive typing required for `MATERIALIZED STREAM VIEW`, we could add a shorthand `CREATE MSV` etc.
```
Comment (mjsax) [inline design-proposals/klip-19-materialize-views.md]: Not sure about this...

File: design-proposals/klip-19-materialize-views.md [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+ * The undefined behaviour if `CREATE TABLE` is used with a topic that is updated by another system
+ * Proper table update semantics for `INSERT VALUES` for non-`VIEW` sources.
+
+Existing demos will need updating. 
+
+Updating existing blogs will be left up to the blog author. 
+
+# Compatibility Implications
+
+The KLIP proposes removing support for `CREATE TABLE AS SELECT`, and potentially `CREATE STREAM AS SELECT`.
```
Comment (mjsax) [inline design-proposals/klip-19-materialize-views.md]: Why removing instead of deprecation?

File: design-proposals/klip-19-materialize-views.md (hunk 1 above)
Comment (big-andy-coates) [reply design-proposals/klip-19-materialize-views.md]: To implement `UPDATE` we need to do an atomic read-update-write of the value. Likely, we'd implement this by materializing the table into rocksdb, which would mean a specific row is owned by a specific KSQL node. Hence any `UPDATE` to that row would need to be routed to that node.

File: design-proposals/klip-19-materialize-views.md (hunk 2 above)
Comment (big-andy-coates) [reply design-proposals/klip-19-materialize-views.md]: Yes, `MATERIALIZED` is pertaining to the data being _materialized_ in a topic.

IMHO, whether a table is also materialized into a state store / rocksdb inside KSQL would ideally be an implementation detail.  The engine may in the future _always_ materialize tables into a rocks db store, or choose to materialize only heavily used tables, etc.  However, crucially, this should be hidden from normal users.  More importantly, I think this is outside the scope of this discussion - though I agree the KLIP should be more clear.

File: design-proposals/klip-19-materialize-views.md (hunk 3 above)
Comment (big-andy-coates) [reply design-proposals/klip-19-materialize-views.md]: This may be true. Looking for peoples thoughts on this. Thanks for your input.

File: design-proposals/klip-19-materialize-views.md (hunk 4 above)
Comment (big-andy-coates) [reply design-proposals/klip-19-materialize-views.md]: It's only an option/possibility.  Could be implemented via CLI's auto complete for example.

File: design-proposals/klip-19-materialize-views.md (hunk 5 above)
Comment (big-andy-coates) [reply design-proposals/klip-19-materialize-views.md]: Because we can make breaking changes at the moment as we're not guaranteeing any backwards compatibility. Of course, we could also just deprecate and remove when we move to v1.0, but that carries more tech debt in the meantime.

File: design-proposals/klip-19-materialize-views.md (hunk 1 above)
Comment (mjsax) [reply design-proposals/klip-19-materialize-views.md]: Well, but don't we have the same issue with `INSERT`? If inserts are async as described, two insert for the same primary key would happen in parallel, but only the first one would be "valid". (`DELETE` would not have this issue as it would be idempotent.)

File: design-proposals/klip-19-materialize-views.md (hunk 4 above)
Comment (mjsax) [reply design-proposals/klip-19-materialize-views.md]: Auto-complete seems a good option -- but I would not make it part of the language.

File: design-proposals/klip-19-materialize-views.md (hunk 5 above)
Comment (mjsax) [reply design-proposals/klip-19-materialize-views.md]: Maybe I am more conservative being used to the Kafka process. But KSQL is around for a few years already and even if "ksqlDB" has a lower version, it has a big user base already. I would always prefer to not break backward compatibility if possible. Just my 2 cents.

**Comment by PeterLindner** [comment]
> I am not sure though atm, if we need MATERIALIZED STREAM VIEW.To me, an output stream is always an independent "copy" of the input data and using ACLs to protect it seems to be sufficient?

Although ACLs are technically sufficient i believe MSVs have some value. The KLIP states in the tl;dr section:

> ACLs should be used for access control, not to enforce correct semantics. Introducing Materialized Views will [...] allow users to express intended mutability of data sources.

IMHO that also holds for streams. KSQL users should not mess with Kafka ACLs to express their intent, right?

File: design-proposals/klip-19-materialize-views.md (hunk 2 above)
Comment (purplefox) [reply design-proposals/klip-19-materialize-views.md]: To avoid confusion on this point, perhaps we should avoid using the term "materialized" internally in the code to mean "state stored in rocksdb" as the new materialized keyword will have a different meaning.

File: design-proposals/klip-19-materialize-views.md [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+### What about streams? 
+
+Streams don't have the complexities of per-key update semantics that tables have. So, it's OK to insert into them, right?
+
+Well, streams are certainly not as clear cut as tables. But maybe there's a case for differentiating between a stream and
+a read-only stream view. 
+
+Consider a stream coming from some upstream system, e.g a click stream from a web-tier. Should a KSQL user be able to 
+add rows to such a stream? Probably not, but the true answer is most likely 'it depends'. It depends on how the organisation
```
Comment (purplefox) [inline design-proposals/klip-19-materialize-views.md]: With the new work on KLIP-15 - creating a new API and client, we are explicitly supporting the use case of inserting rows into streams. We want apps (including web apps) to use the new ksqlDB clients (or use the new HTTP2 API directly) to write apps including "web tiers". This is a key point for a great user experience as we don't want apps to also have to manage a Kafka client to insert events into their click stream.
Some users will continue to use Kafka directly for their click stream but I don't think we should discourage users from the new way which we're strategically targeting. (I.e. "Probably not" seems too strong) ;)

File: design-proposals/klip-19-materialize-views.md [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+It can be argued that using ACLs are a good fit for controlling if a KSQL user can insert into a stream. There's nothing
+semantically wrong with allow the insert: a stream is just an append-only list of immutable 'facts'. Appending one more 
+fact does not change the previous or following facts. However, supporting read-only stream views may provide an easier 
+mental model for users to work with. If a user does a `CREATE MATERIALIZED STREAM VIEW AS SELECT` rather than  
+`CREATE STREAM AS SELECT` then it will be clear KSQL does not allow `INSERT`s into the stream.
+
+As a discussion point, and as a mirror of the proposed `MATERIALIZED VIEW` changes above, we propose introducing a 
+`CREATE MATERIALIZED STREAM VIEW` statement to allow users to import stream datasets as read-only.
+
+The existing `CREATE STREAM` statement will be for defining new streams or importing an existing topic as a stream, not a view.
```
Comment (purplefox) [inline design-proposals/klip-19-materialize-views.md]: Will this support the user creating a stream in KSQL that is both a) Owned by KSQL and b) Allows inserts?

File: design-proposals/klip-19-materialize-views.md (hunk 3 above)
Comment (purplefox) [reply design-proposals/klip-19-materialize-views.md]: I'd favour removing it unless it's something that's going to be useful to most users. Smaller, simpler API = less user confusion = better user experience = more adoption :)

File: design-proposals/klip-19-materialize-views.md (hunk 4 above)
Comment (purplefox) [reply design-proposals/klip-19-materialize-views.md]: These kinds of things won't be typed very often so optimising for the time taken here doesn't make a lot of sense imo.

**Comment by big-andy-coates** [comment]
@mjsax 

> I think a proper "ownership model" for ksqlDB is quite important. I am not sure though atm, if we need MATERIALIZED STREAM VIEW.To me, an output stream is always an independent "copy" of the input data and using ACLs to protect it seems to be sufficient?

I think our the ksql model should be semantically correct _without_  ACLs.  As I say at the top of the KLIP:

> However, ACLs should be used for access control, not to enforce correct semantics

Many people may be running against Kafka clusters without ACLs enabled.

File: design-proposals/klip-19-materialize-views.md (hunk 1 above)
Comment (big-andy-coates) [reply design-proposals/klip-19-materialize-views.md]: The `INSERT` is more correctly an `UPSERT` (We should probably correct that syntax too).  Each `INSERT` fully replaces any previous value.

However, an `UPDATE` may only provide partial information i.e. `UPDATE foo SET col1 = 1 WHERE col2 = 3;`. This requires atomic read-update-writes.

I've added more details to the KLIP.

File: design-proposals/klip-19-materialize-views.md (hunk 3 above)
Comment (big-andy-coates) [reply design-proposals/klip-19-materialize-views.md]: Great, so one vote either way. Perfect! ;) :p

File: design-proposals/klip-19-materialize-views.md (hunk 4 above)
Comment (big-andy-coates) [reply design-proposals/klip-19-materialize-views.md]: removed!

File: design-proposals/klip-19-materialize-views.md (hunk 6 above)
Comment (big-andy-coates) [reply design-proposals/klip-19-materialize-views.md]: Yep agree. It does go on to explain 'it depends'. But I've reworded slightly.

File: design-proposals/klip-19-materialize-views.md (hunk 7 above)
Comment (big-andy-coates) [reply design-proposals/klip-19-materialize-views.md]: Yes. This is explained later in the doc.

File: design-proposals/klip-19-materialize-views.md (hunk 3 above)
Comment (PeterLindner) [reply design-proposals/klip-19-materialize-views.md]: If `CSAS` is removed one would have to create a new 'insertable' stream and `UNION` it with the old stream to achieve the same result which would require 2 new streams instead of one, right?
For performance reasons it may be beneficial to keep `CSAS`, but you can judge that way better than I could.

Comment (guozhangwang) [reply design-proposals/klip-19-materialize-views.md:171]: I feel that we only need to distinguish "creating a stream / table" from "creating a read-only stream / table" here. But instead of making the newly introduced keyword as "opt-in" (i.e. without the keyword the stream / table would be mutable) maybe it's better making them "opt-out" (i.e. by default they are immutable). The rationale I have is that in practice, most streams / tables would actually be immutable, and perhaps only the source streams need to be mutable. And since not many people have been using `INSERT INTO` in practice so maybe it's okay to break this semantic compatibility:

```
CREATE STREAM AS SELECT   // create an un-insertable stream
CREATE TABLE AS SELECT   // create an un-insertable table
CREATE MUTABLE STREAM AS SELECT     // create un-insertable stream
CREATE MUTABLE TABLE AS SELECT     // create un-insertable table
```

Admittedly, and for tables the `immutable` keyword has a more commonly used alternative `materialized view`, so if we decide to go with `opt-in` I'm also fine.

Comment (guozhangwang) [reply design-proposals/klip-19-materialize-views.md:171]: Also a key question we've not yet addressed is, if we ever wants to make ksqlDB to be a real RDBMS, because otherwise we could say we exclude supporting mutable tables at all (i.e. there's only materialized views, in which KLIP-11's proposal seems more reasonable to me).

If we decided to still support direct-inserts-into-table in the long run, I think we should still make this distinguishment --- whatever calling it materialized views or immutable or read-only --- as early as possible.

Comment (mjsax) [reply design-proposals/klip-19-materialize-views.md:171]: > And since not many people have been using INSERT INTO in practice

How do you/we know this?

About your proposal to flip the default from mutable to immutable: I am not a fan of it, because I think we should try to stick with standard SQL if possible. Additionally, if we remove CT syntax completely, and only have CMV syntax, it's pretty clear to users what going on.

Also note that `INSERT INTO table` is already supported. It's implemented as an append to the corresponding topic (ie, it's an async write; there is no read-after-write consistency). I assume that this feature would be "dropped" if we switch all statements from CT to CMV? \cc @big-andy-coates

Comment (guozhangwang) [reply design-proposals/klip-19-materialize-views.md:171]: Yeah that's a fair point. My rationale is just that: in the future do we expect majority of TABLEs to be immutable or mutable, and the language design should be best making the majority case with less keywords if we are designing it for the first time from scratch.

My bet is that immutable tables would be majority, and hence they should enjoy the CT statement with less keywords while the mutable CT should use more keywords. Of course, that's assuming we are re-designing the language from scratch but that's not the case, today effectively mutable CT uses less words, so the question is whether it's worth-while to break this semantic compatibility for the sake of the above argument: if not too many people have already used `INSERT INTO` maybe it's worthwhile, but admittedly I'm do not know statistically how many are there indeed.

Comment (mjsax) [reply design-proposals/klip-19-materialize-views.md:171]: Well, SQL has already a statement to create immutable tables: `CREATE MATERIALIZE VIEW` -- thus, IMHO we should just use what is already there. Why "flipping" it around introduce non-standard SQL while SQL offers everything we need already?
```
// standard SQL
CREATE TABLE... // mutable
CREATE MATERIALIZED VIEW... // immutable

// why is this better?
CREATE MUTABLE TABLE... // mutable
CREATE TABLE... // immutable
```

If we assume that most people want/need immutable tables, it seems sufficient to educate them to use CMV statements. No need to introduce new syntax and no need to change the semantics of the well established CT statement (I would expect that this would be rather confusing to users).

Comment (guozhangwang) [reply design-proposals/klip-19-materialize-views.md:171]: Yeah I think the existing syntax in SQL is a strong one compared to merely saving some keywords, I'm happy to fold my suggestion about flipping it
